Python: examples, and a Recurrence class.

rustyrussell opened the pull request.

rustyrussell committed:
Makefile: `python` is python2 on Ubuntu 20.4.2 LTS

But python3 does not exist on Windows.  So allow overriding.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
Makefile: add checker for quotes, general check target.

Stolen from lnprototest, but really should be in lightning-rfc repo.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
GitHub: Add CI runners for pull requests and main branch.

It used a virtualenv, so avoid hardcoding /usr/bin/python3.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
python: implement Recurrence class.

This is very non-trivial!  Periods require advanced time handling as
the spacing is not a simple seconds, but can be months or years.

Fortunately, there are test vectors!

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
examples: provide a simple decode example.

And fix up the case where we would raise an exception on truncated
bech32.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
Python: make 'flake8' linter happy.

Mainly whitespace changes to generator, but it found one real bug: we
were using [fieldname] instead of ["fieldname"] for length fields in
both Python and Javascript.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
python: make mypy happy.

Mainly fixing type annotations, but also changed table generation to
use Tuples, not Lists, which makes more sense since they are constant.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
Fix package name for pyln.proto

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
TODO: more TODOs for packages.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

adi2011 commented on Makefile:
```diff
@@ -1,4 +1,21 @@
+
+check-quotes/%.js: %.js
+	@echo $@
+	@tools/check_quotes.py --comment-start='// ' --comment-continue='//' --boltdir=lightning-rfc/ $<
```
should we add '/* .. */' for long comments?

rustyrussell replied:
Currently you have to define a single format; as long as we're consistent with how we quote the BOLTs, I guess this is OK?

adi2011 commented on python/bolt12.py:
```diff
@@ -111,24 +117,174 @@ def simple_bech32_decode(bech32str: str) -> Tuple[str, bytes]:
         return None, None
```
if sep == -1, it should throw Error(Separator not present) instead of 'return None, None', that makes it for the user to debug maybe?

rustyrussell replied:
Yes, it would be nice to give the user more information, but Python exceptions are an awkward way to do it unless you create a new class (imagine a GUI app, which wanted to highlight where the error was).

Hmm, maybe you're right...

adi2011 commented on python/bolt12.py:
```diff
@@ -111,24 +117,174 @@ def simple_bech32_decode(bech32str: str) -> Tuple[str, bytes]:
-    return hrp, bytes(bech32.convertbits(ret5, 5, 8, False))
+
+    intarr = bech32.convertbits(ret5, 5, 8, False)
+    if intarr is None:
```
Okay will it ever be None? I haven't handled it in JS..

rustyrussell replied:
No, it could only return None if we gave it a value which wasn't in range 0-31. Obviously we don't, since we just decoded it.  But mypy complains it can be None, so I handled it :)

adi2011 commented on python/bolt12.py:
```diff
@@ -111,24 +117,174 @@ def simple_bech32_decode(bech32str: str) -> Tuple[str, bytes]:
+    for c in bech32str[sep + 1:]:
         pos = bech32.CHARSET.find(c)
         if pos == -1:
             return None, None
```
Here as well, it should throw an error with the invalid-letter?

rustyrussell committed:
python/bolt12: make simple_bech32_decode raise an exception

This is probably more informative than returning None, None.

Suggested-by: @adi2011
Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>
